cmp/ag-grid/AgGridModel.js

```diff
@@ -59,6 +63,198 @@ export class AgGridModel {
+     *
+     * Some of the state may be data-dependent. Specifically the expandState and filterState. It is
+     * recommended that applications wait until the data has been loaded in the grid before setting
+     * the state if including those elements. This method can be called immediately after the data
+     * has been loaded via agApi.setRowData
+     *
+     * @param {AgGridState} state
+     */
+    setState(state) {
+        const {columnState, sortState, expandState, filterState, miscState} = state;
+        if (columnState) this.setColumnState(columnState);
+        if (sortState) this.setSortState(sortState);
+        if (expandState) this.setExpandState(expandState);
+        if (filterState) this.setFilterState(filterState);
+        if (miscState) this.setMiscState(miscState);
+    }
+
+    /**
+     * @returns {AgGridMiscState}
+     */
+    getMiscState() {
+        return {
+            panelId: this.agApi.getOpenedToolPanel()
+        };
+    }
+
+    /**
+     * Sets the grid state which doesn't fit into the other buckets.
+     * @param {AgGridMiscState} miscState
+     */
+    setMiscState(miscState) {
+        const {agApi} = this,
+            {panelId} = miscState;
+
+        if (isNil(panelId)) {
+            agApi.closeToolPanel();
+        } else {
+            agApi.openToolPanel(panelId);
+        }
+    }
+
+    /**
+     * @returns {Object[]} - current filter state of the grid. @see https://www.ag-grid.com/javascript-grid-filtering/#get-set-all-filter-models
+     */
+    getFilterState() {
+        const {agApi} = this;
+        return agApi.getFilterModel();
+    }
+
+    /**
+     * Sets the grid filter state. Note that this state may be data-dependent, depending on the types
+     * of filter being used.
+     *
+     * @param {Object[]} filterState
+     */
+    setFilterState(filterState) {
+        const {agApi} = this;
+
+        agApi.setFilterModel(filterState);
+        agApi.onFilterChanged();
+    }
+
+    /**
+     * @returns {Object[]} - current sort state of the grid. @see https://www.ag-grid.com/javascript-grid-sorting/#sorting-api
+     */
+    getSortState() {
+        const {agApi, agColumnApi} = this,
+            sortModel = agApi.getSortModel(),
+            isPivot = agColumnApi.isPivotMode();
+
+        // When we have pivot columns we need to make sure we store the path to the sorted column
+        // using the pivot keys and value column id, instead of using the auto-generate secondary
+        // column id as this could be different with different data or even with the same data based
+        // on the state of the grid when pivot mode was turned on
+        if (isPivot && !isEmpty(agColumnApi.getPivotColumns())) {
+            const secondaryCols = agColumnApi.getSecondaryColumns();
+
+            sortModel.forEach(sort => {
+                const col = secondaryCols.find(it => it.colId === sort.colId);
+                sort.colId = this.getPivotColumnId(col);
```

Comment:
I was able to trigger an NPE here when attempting to save state on the Toolbox example. 

* Start in non-pivot mode
* Sort on any column - I clicked sector
* Log grid state - sector col in sortModel/sortState
* Enable pivot, do not sort anything.
* Attempt to log or save grid state - throws. The sortModel still contains sector but we fail to find a column for it, so col is undefined and we throw in `getPivotColumnId()`.

Did not investigate further - assume we should remove that column from the returned sortModel array entirely?

In general, wondering if we should attempt to "safely" get/set each chunk of state - i.e. wrap in a try/catch? Of course I don't want to paper over errors we can investigate and better handle in the code, but if we develop a sense that certain unexpected pivot / data / column conditions might change in ways that are going to remain difficult to predict, we might want to make this mechanism as robust as possible in terms of doing what it can, vs throwing and not being able to do anything.

Reply:
This is an interesting case, and probably something we will continue running into, where ag-Grid doesn't really do a great job of maintaining it's internal state models when switching between the 3 possible modes (non-pivot, pivot with no pivot cols, pivot with pivot cols).

I think you are correct that in this case we just want to remove that entry from the sortModel, it is not actually representative of the current state of the grid that the user is seeing. One could argue that this is actually a bug in ag-Grid.

I think it is a good idea to make retrieval of each chunk of state "safe", but would we want the application to be able to throw up a warning toast that not all requested grid state was able to be captured? Should be include an `errors` prop in the returned state object for app usage? (and also would probably be useful for debugging the state json )